### ISSUE TITLE
Display corruption in the observer

### ISSUE THREAD

**Issue by galdor** [erlang/otp#4812]
# Issue
The problem can be reproduced every time, even just running `erl`:

- Start the observer.
- Select the "Applications" tab.
- Click on any of the box representing a process.
- Select another tab.
- Go back to the "Applications" tab.
- The content of the other tab is still visible, with the content of the "Applications" tab displayed over it.

At this point, selecting other tabs and going back, or moving any other window over the observer window, will cause various other display bugs. The situation cannot be recovered without stopping the observer and starting it again.

The following log message is printed when the observer starts:
```
(Erlang:55743): Gtk-WARNING **: 07:31:09.109: Drawing a gadget with negative dimensions. Did you forget to allocate a size? (node tab owner GtkNotebook)
```

However it always appears when the observer starts, before the bug is triggered, so it is hard to know if it is related.

# Screenshot
![observer-display-bug](https://user-images.githubusercontent.com/37804/117528087-b9313400-afd0-11eb-85f3-4b4a2bdbfd42.png)

# Environment
- Linux (Archlinux).
- Erlang/OTP 23.3.2.
- wxGTK 3.0.5.1.
- i3 4.19.1.

**Comment by dgud** [comment]
Can't reproduce, it works fine on ubuntu-18 and ubuntu-20.

**Comment by galdor** [comment]
I just updated to OTP 24, and the problem is still exactly the same.

**Comment by paulo-ferraz-oliveira** [comment]
I can't replicate this. I'm on OTP 24.0, on MacOSX.

I do get `2021-05-13 11:28:01.230 beam.smp[76044:24317016] IMKClient Stall detected, *please Report* your user scenario attaching a spindump (or sysdiagnose) that captures the problem - (imkxpc_attributesForCharacterIndex:reply:) block performed very slowly (1.35 secs).` in the shell, though, but it's probably unrelated.

@galdor, might this be a Linux-specific issue or a "wxGTK for Linux" one?

**Comment by galdor** [comment]
It is probably a combination of factors, from the gtk and/or wxgtk version to the way they are used by Erlang. But I have no issue whatsoever with the multiple GTK-based applications I run on my system for years.

Let me know if there is something I can do to help debug this mess.

**Comment by lhoguin** [comment]
What's the GTK version?

The first thing that would help is figure out which component is doing something wrong. Does it happen on OTP 22/24? Does it happen on wxGTK 2? Does it happen with a different GTK version?

**Comment by galdor** [comment]
Issue updated for the GTK version, 3.24.29. As I said, it happens at least for 23.x and 24.

**Comment by dgud** [comment]
I don't think this is a bug in observer nor in wx, so I don't believe the problem is ours, either wxWidgets or a gtk bug.
I'm closing the issue.

**Comment by galdor** [comment]
> I don't think this is a bug in observer nor in wx
Given that I never had any display corruption of any kind with any of the dozens of GTK applications I run for ~15 years on Linux, and that the observer continues to exhibit the same issue version after version, I would be inclined to blame either observer or wx.

I will just stop using observer and write my own visualization interface.

**Comment by lhoguin** [comment]
Do you have GTK/GDK/other environment variables influencing the GTK behavior? Are you perhaps using the OBSERVER_SCALE environment variable?

**Comment by galdor** [comment]
No environment variable for GTK/GDK, no `OBSERVER_SCALE`. And I have observed (pun not indented) the exact same corruption issue on multiple computers for quite some time.

It is no big deal, I intended to write a web -based visualization interface because the observer is not the most practical interface anyway. It is just sad that is it yet one more component slowly rotting in its corner.

**Comment by lhoguin** [comment]
Could it be related to the theme you're using? On my machine Observer has a gray background. On your screenshot I can see that it is white. Could it be something in the theme's CSS causing Observer to misbehave? Have you tried switching the GTK theme to a completely different one to see if the problem still exists?

**Comment by galdor** [comment]
I just tried with a few other GTK3 themes, including Adwaita, Arc and Adapta, and I can confirm the exact same corruption issue appears.

**Comment by lhoguin** [comment]
Yeah I don't know. I don't have i3 but it's the last remaining thing I can think of. Maybe the tiling changes the window properties in a way that Observer doesn't handle properly when it does the drawing of its custom components.

**Comment by dgud** [comment]
What happens if you resize the window (i.e. force a redraw) when you see the display corruption?

**Comment by galdor** [comment]
Resizing the window causes a ton more display artifacts. It really look like the background of the canvas is never redrawed. 

Shrinking the window a bit too much (less than half of the original widths causes dozens of extra GTK warnings such as:

```
(Erlang:88937): Gtk-CRITICAL **: 11:25:40.867: gtk_box_gadget_distribute: assertion 'size >= 0' failed in GtkScrollbar
```

and

```
(Erlang:88937): Gtk-WARNING **: 11:25:55.869: Negative content width -25 (allocation 1, extents 13x13) while allocating gadget (node button, owner GtkButton)
```

**Comment by dgud** [comment]
Sounds/Seems like a wxWidgets-gtk combo,  can you try compiling wxWidgets your self, 
i.e. don't use the distro version. 
Maybe test with wxWidgets-3.1.5 
  ( if you are using OTP < 24 you must enable --enable-compat28, when running configure for wxWidgets).
